source/azure_iot_json_reader.c

```
@@ -0,0 +1,258 @@
+    az_result xCoreResult;
+    az_span xJSONSpan;
+
+    if( ( pxReader == NULL ) || ( pucBuffer == NULL ) || ( ulBufferSize == 0 ) )
+    {
+        AZLogError( ( "AzureIoTJSONReader_Init failed: invalid argument" ) );
+        xResult = eAzureIoTErrorInvalidArgument;
+    }
+    else
+    {
+        xJSONSpan = az_span_create( ( uint8_t * ) pucBuffer, ( int32_t ) ulBufferSize );
+
+        if( az_result_failed( xCoreResult = az_json_reader_init( &pxReader->_internal.xCoreReader, xJSONSpan, NULL ) ) )
+        {
+            AZLogError( ( "Could not initialize the JSON reader: core error=0x%08x", xCoreResult ) );
+            xResult = eAzureIoTErrorFailed;
+        }
+        else
+        {
+            xResult = eAzureIoTSuccess;
+        }
+    }
+
+    return xResult;
+}
+
+
+AzureIoTResult_t AzureIoTJSONReader_NextToken( AzureIoTJSONReader_t * pxReader )
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+
+    if( pxReader == NULL )
+    {
+        AZLogError( ( "AzureIoTJSONReader_NextToken failed: invalid argument" ) );
+        xResult = eAzureIoTErrorInvalidArgument;
+    }
+    else
+    {
+        if( az_result_failed( xCoreResult = az_json_reader_next_token( &pxReader->_internal.xCoreReader ) ) )
+        {
+            AZLogError( ( "Could not get next JSON token: core error=0x%08x", xCoreResult ) );
+            xResult = eAzureIoTErrorFailed;
+        }
+        else
+        {
+            xResult = eAzureIoTSuccess;
+        }
+    }
+
+    return xResult;
+}
+
+AzureIoTResult_t AzureIoTJSONReader_SkipChildren( AzureIoTJSONReader_t * pxReader )
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+
+    if( pxReader == NULL )
+    {
+        AZLogError( ( "AzureIoTJSONReader_SkipChildren failed: invalid argument" ) );
+        xResult = eAzureIoTErrorInvalidArgument;
+    }
+    else
+    {
+        if( az_result_failed( xCoreResult = az_json_reader_skip_children( &pxReader->_internal.xCoreReader ) ) )
```

should we translate error from Embedded sdk to Middleware layer, as currently we are masking all error to eAzureIoTErrorFailed

---

how about a work item to have something like an `include/internal/az_result_translate.h`  with a function which translates relevant errors from embedded to middleware. Seems like something that would be beneficial to all our middleware api's and is beyond the scope of this PR.

---

https://github.com/Azure/azure-iot-middleware-freertos/issues/137